### ISSUE TITLE
Automatic Differentiation

### ISSUE THREAD

**Issue by hmustafajr** [hmustafajr/PyTorch-Fundamentals#5]
**Automatic Differentiation with torch.autograd**
When training neural networks, the most frequently used algorithm is **back propagation**. In this algorithm, parameters (model weights) are adjusted according to the **gradient** of the loss function with respect to the given parameter. The loss function calculates the difference between the expected output and the actual output that a neural network produces. The goal is to get the result of the loss function as close to zero as possible. The algorithm traverses backwards through the neural network to adjust the weights and bias to retrain the model. That's why it's called back propagation. This back and forward process of retraining the model over time to reduce the loss to 0 is called the gradient descent.

To complete those gradients. PyTorch has a built-in differentiation engine called torch.autograd. It supports automatic computation of gradient for any computational graph.

Consider the simplest one-layer neural network, with input **x**, parameters **w** and **b**, and some loss function. It can be defined in PyTorch in the following manner:

[Torch Autograd.py](https://github.com/hmustafajr/Python-Toolbox/blob/main/Torch%20Autograd.py)

**Computing Gradients**
To optimize weights of parameters in the neural network, we need to compute the derivatives of our loss function with respect to parameters, namely, we need   ∂loss / ∂w  and  ∂loss / ∂w under some fixed values of **x** and **y**. To compute those derivatives, we call **loss,backward()**, and then retrieve the values from **w.grad** and **b.grad**.

**Note**: We can only obtain the _grad_ properties for the leaf nodes of the computational graph that have the _requires_grad_ property set to _True_. For all other nodes in our graph, gradients will are not available. In addition, we can only perform gradient calculations using _backward_ once on a given graph, for performance reasons. If we need to perform several _backward_ calls on the same graph, we need to pass _retain_graph=True_ to the _backward_ call.

​[Computing Gradients](https://github.com/hmustafajr/Python-Toolbox/blob/main/Computing%20Gradients.py)
 
**Disabling Gradient Tracking**
​By default, all tensors with **requires_grad=True** are tracking their computational history and support gradient computation. However, there are some cases when we don't need to do that, for example, when we've trained the model and just want to apply it to some input data; that is we only want to do **forward** computations through the network. We can stop tracking computations by surrounding our computation code with a **torch.no_grad()** block.

[Disable Gradient Tracking](https://github.com/hmustafajr/Python-Toolbox/blob/main/Disable%20Gradient.py)

Another method is to use **detach()** on the tensor:
```python
z = torch.matmul(x,w)+b
z_det = z.detach()
print(z_det.requires_grad)
``` 

There are a couple of reasons that anyone would want to disable gradient tracking.
- Marking some parameters in your neural network as **frozen parameters**. This is quite common in scenarios where fine-tuning a pretrained network.
- To **speed up computations** when you are only doing a forward pass, because computations on tensors that don't track gradients are more efficient.
